src/validators/union.rs

```diff
@@ -134,16 +137,42 @@ impl UnionValidator {
                     _ => {
                         // success should always have an exactness
                         debug_assert_ne!(state.exactness, None);
+
                         let new_exactness = state.exactness.unwrap_or(Exactness::Lax);
-                        // if the new result has higher exactness than the current success, replace it
-                        if success
-                            .as_ref()
-                            .map_or(true, |(_, current_exactness)| *current_exactness < new_exactness)
-                        {
-                            // TODO: is there a possible optimization here, where once there has
-                            // been one success, we turn on strict mode, to avoid unnecessary
-                            // coercions for further validation?
-                            success = Some((new_success, new_exactness));
+
+                        // in the case where the exactness is Strict or Lax, we also check
+                        // the number of fields in the model, and prefer the one with more fields
+                        // which is useful for choosing a more specific model in a union, like a subclass
+                        let new_num_fields: Option<usize> = match choice {
```

Comment:
One solution to this is to have `num_fields` be part of the `Validator` trait, and things like function validators could then reach into their inner validators. But that's then coupling the `Validator` trait to very specific details of the union validation, so it's not super fun either.

Also, what happens for weird cases here like `list[ModelA] | list[ModelB]`? That's already terrible performance, but should we strive to make it work like `ModelA | ModelB`? Probably, yes. Putting a method on the `Validator` trait so that all validators get to say what their "num fields" is might be best.

Reply:
> One solution to this is to have num_fields be part of the Validator trait, and things like function validators could then reach into their inner validators.

Nice. I like this approach for now, and it's pretty easy for us to change in the future bc it's an internal detail. I've added a default implementation so that you have to explicitly implement the `num_fields` method on a validator if you'd like to customize the metric.

> Also, what happens for weird cases here like list[ModelA] | list[ModelB]? That's already terrible performance, but should we strive to make it work like ModelA | ModelB? Probably, yes.

Good question. This PR doesn't address this case:

```py
from pydantic import BaseModel, TypeAdapter

class ModelA(BaseModel):
    a: str

class ModelB(ModelA):
    b: str

ta = TypeAdapter(list[ModelA] | list[ModelB])

# this is not right, imo
print(repr(ta.validate_python([{'a': 'a', 'b': 'b'}])))
#> [ModelA(a='a')

# this is right
print(repr(ta.validate_python([{'a': 'a'}, {'a': 'a', 'b': 'b'}])))
#> [ModelA(a='a'), ModelA(a='a')]

# this is not right, imo
print(repr(ta.validate_python([{'a': 'a', 'b': 'b'}, {'a': 'a', 'b': 'b'}])))
#> [ModelA(a='a'), ModelA(a='a')]
```

That being said, I do think this is a good step forward in the smart union validation logic, and I think we should strive to fix these other cases above within the framework we're starting to use in this PR.